### PR TITLE
[Model] New model support for Motif-1-Tiny

### PR DESCRIPTION
**Pull request by ca1207**
## Purpose
New model for : 
https://huggingface.co/Motif-Technologies/Motif-2.6B
https://huggingface.co/Motif-Technologies/Motif-2.6b-v1.1-LC
Tech report (https://arxiv.org/pdf/2508.09148)

Implemented the Polynorm custom kernel based on https://arxiv.org/abs/2411.03884
co-author : @WyldeCat 
## Test Plan
benchmark for Polynorm kernel (benchmarks/kernels/benchmark_polynorm.py)
## Test Result
Benchmark Results

<details>
<summary>Results for dim = 2048</summary>

| dim  | batch_size | seq_len | Naive   | vLLM   | speed up |
|------|------------|---------|---------|--------|----------|
| 2048 | 1          | 64      | 197.76  | 8.61   | 22.97    |
| 2048 | 1          | 128     | 197.31  | 9.18   | 21.48    |
| 2048 | 1          | 256     | 194.34  | 11.36  | 17.11    |
| 2048 | 1          | 512     | 274.91  | 13.66  | 20.12    |
| 2048 | 1          | 1024    | 552.77  | 20.51  | 26.95    |
| 2048 | 4          | 64      | 194.46  | 11.62  | 16.74    |
| 2048 | 4          | 128     | 275.07  | 13.63  | 20.18    |
| 2048 | 4          | 256     | 551.94  | 20.54  | 26.87    |
| 2048 | 4          | 512     | 1044.08 | 34.53  | 30.24    |
| 2048 | 4          | 1024    | 1971.10 | 61.57  | 32.02    |
| 2048 | 16         | 64      | 552.70  | 20.58  | 26.86    |
| 2048 | 16         | 128     | 1045.07 | 34.56  | 30.24    |
| 2048 | 16         | 256     | 1971.33 | 61.15  | 32.24    |
| 2048 | 16         | 512     | 3811.10 | 129.89 | 29.34    |
| 2048 | 16         | 1024    | 7492.06 | 237.12 | 31.60    |
| 2048 | 64         | 64      | 1971.89 | 61.54  | 32.04    |
| 2048 | 64         | 128     | 3811.01 | 129.63 | 29.40    |
| 2048 | 64         | 256     | 7491.81 | 236.13 | 31.73    |
| 2048 | 64         | 512     | 14854.58| 479.70 | 30.97    |
| 2048 | 64         | 1024    | 29569.15| 975.44 | 30.31    |

</details>

<details>
<summary>Results for dim = 4096</summary>

| dim  | batch_size | seq_len | Naive   | vLLM   | speed up |
|------|------------|---------|---------|--------|----------|
| 4096 | 1          | 64      | 196.59  | 9.86   | 19.95    |
| 4096 | 1          | 128     | 196.13  | 10.78  | 18.19    |
| 4096 | 1          | 256     | 269.94  | 14.40  | 18.75    |
| 4096 | 1          | 512     | 545.79  | 21.22  | 25.73    |
| 4096 | 1          | 1024    | 1034.14 | 38.98  | 26.53    |
| 4096 | 4          | 64      | 269.34  | 14.50  | 18.58    |
| 4096 | 4          | 128     | 545.15  | 21.09  | 25.85    |
| 4096 | 4          | 256     | 1033.57 | 39.01  | 26.50    |
| 4096 | 4          | 512     | 1963.90 | 71.52  | 27.46    |
| 4096 | 4          | 1024    | 3804.32 | 135.81 | 28.01    |
| 4096 | 16         | 64      | 1034.14 | 39.09  | 26.46    |
| 4096 | 16         | 128     | 1964.13 | 71.49  | 27.47    |
| 4096 | 16         | 256     | 3804.16 | 135.81 | 28.01    |
| 4096 | 16         | 512     | 7476.80 | 265.06 | 28.21    |
| 4096 | 16         | 1024    | 14831.06| 523.97 | 28.31    |
| 4096 | 64         | 64      | 3803.46 | 135.74 | 28.02    |
| 4096 | 64         | 128     | 7477.31 | 265.09 | 28.21    |
| 4096 | 64         | 256     | 14832.98| 524.03 | 28.31    |
| 4096 | 64         | 512     | 29530.88| 1041.42| 28.36    |
| 4096 | 64         | 1024    | 58956.83| 2080.24| 28.34    |

</details>

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/model_executor/models/motif.py [hunk 1]
```diff
@@ -0,0 +1,558 @@
...
+@support_torch_compile(
+    dynamic_arg_dims={
+        "input_ids": 0,
+        # positions is of shape (3, seq_len) if mrope is enabled for qwen2-vl,
+        # otherwise (seq_len, ).
+        "positions": -1,
+        "intermediate_tensors": 0,
+        "inputs_embeds": 0,
+    })
+class MotifModel(nn.Module):
```
Comment (jeejeelee) [inline vllm/model_executor/models/motif.py]: Are there differences between this model and LLaMA? Can we inherit from LLaMA, similar to: https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/telechat2.py

File: docs/models/supported_models.md [hunk 2]
```diff
@@ -382,6 +382,7 @@ th {
 | `MiniCPM3ForCausalLM` | MiniCPM3 | `openbmb/MiniCPM3-4B`, etc. | ✅︎ | ✅︎ | ✅︎ |
 | `MistralForCausalLM` | Mistral, Mistral-Instruct | `mistralai/Mistral-7B-v0.1`, `mistralai/Mistral-7B-Instruct-v0.1`, etc. | ✅︎ | ✅︎ | ✅︎ |
 | `MixtralForCausalLM` | Mixtral-8x7B, Mixtral-8x7B-Instruct | `mistralai/Mixtral-8x7B-v0.1`, `mistralai/Mixtral-8x7B-Instruct-v0.1`, `mistral-community/Mixtral-8x22B-v0.1`, etc. | ✅︎ | ✅︎ | ✅︎ |
+| `MotifForCausalLM` | Motif-1-Tiny | `Motif-Technologies/Motif-2.6B`, `Motif-Technologies/Motif-2.6b-v1.1-LC`, etc. | | ✅︎ | |
```
Comment (jeejeelee) [inline docs/models/supported_models.md]:
```suggestion
| `MotifForCausalLM` | Motif-1-Tiny | `Motif-Technologies/Motif-2.6B`, `Motif-Technologies/Motif-2.6b-v1.1-LC`, etc. | ✅︎ | ✅︎ | |
```

File: tests/models/registry.py [hunk 3]
```diff
@@ -257,6 +257,9 @@ def check_available_online(
     "MixtralForCausalLM": _HfExamplesInfo("mistralai/Mixtral-8x7B-Instruct-v0.1",  # noqa: E501
                                           {"tiny": "TitanML/tiny-mixtral"}),  # noqa: E501
     "QuantMixtralForCausalLM": _HfExamplesInfo("mistral-community/Mixtral-8x22B-v0.1-AWQ"),  # noqa: E501
+    "MotifForCausalLM": _HfExamplesInfo("Motif-Technologies/Motif-2.6B",
+                                        trust_remote_code=True,
+                                        v0_only=True),
```
Comment (jeejeelee) [inline tests/models/registry.py:290]: QQ: Why does this model only support V0?

Comment (ca1207) [reply tests/models/registry.py:290]: Our model uses the differential attention introduced in this [PR](https://github.com/vllm-project/vllm/pull/20702).
As you can see [here,](https://github.com/vllm-project/vllm/pull/20702/files#r2196997147) the current differential attention backend does not support chunked prefill, so it is only available in v0.

File: vllm/model_executor/models/motif.py (hunk 1 above)
Comment (ca1207) [reply vllm/model_executor/models/motif.py]: The key differences between our model and LLaMA are as follows:
we employ differential attention in the attention mechanism,
use Polynorm as the activation function,
and apply different epsilon values for RMSNorm between the attention layers and the other layers.

Given these differences , it may be more appropriate not to inherit directly from LLaMA.

Please refer to [this paper](https://arxiv.org/pdf/2410.05258) for details on differential attention, and to [this paper](https://arxiv.org/abs/2411.03884) for further information on PolyNorm.

File: docs/models/supported_models.md (hunk 2 above)
Comment (ca1207) [reply docs/models/supported_models.md]: Could you please elaborate a bit more on this comment?
We have not tested with LoRA, so could you clarify which aspects led you to conclude that LoRA would be feasible?

**Comment by WyldeCat** [comment]
Hi @jeejeelee , just a gentle ping on this PR 🙂.
Would love your feedback when you get a chance.

File: vllm/model_executor/models/motif.py (hunk 1 above)
Comment (jeejeelee) [reply vllm/model_executor/models/motif.py]: It looks like the difference is mainly in the `MotifDecoderLayer`. If this is the case, we can also refer to [telechat2](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/telechat2.py) implementation  to reduce some code, such as PP and weight load

File: docs/models/supported_models.md (hunk 2 above)
Comment (jeejeelee) [reply docs/models/supported_models.md]: In your implementation, `MotifForCausalLM` inherits  from `SupportsLoRA`. 
You can also set `enable_lora` to test if the model can be initialized successfully to verify this

File: tests/models/registry.py (hunk 3 above)
Comment (jeejeelee) [reply tests/models/registry.py:290]: @DarkLight1337  If this model  only supports V0, how should we handle this now?

Comment (DarkLight1337) [reply tests/models/registry.py:290]: You can have the model inherit from `SupportsV0Only` explicitly

Comment (DarkLight1337) [reply tests/models/registry.py:290]: And state the reason why via code comment for future reference

Comment (DarkLight1337) [reply tests/models/registry.py:290]: @WoosukKwon is it possible to turn off chunked prefill for generative models in V1?

**Comment by ca1207** [comment]
@jeejeelee I have addressed your PR review comments. Please feel free to take a look whenever you have some time.

**Comment by hmellor** [comment]
- Please merge from main to hopefully fix the failing test
- Please fix `pre-commit`

File: vllm/model_executor/models/motif.py [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+            cache_config=cache_config,
+            prefix=f"{prefix}.self_attn",
+            attn_type=attn_type,
+        )
+        self.mlp = MotifMLP(
+            hidden_size=self.hidden_size,
+            intermediate_size=config.intermediate_size,
+            hidden_act=config.hidden_act,
+            quant_config=quant_config,
+            bias=getattr(config, "mlp_bias", False),
```
Comment (jeejeelee) [inline vllm/model_executor/models/motif.py]: By learning from https://huggingface.co/Motif-Technologies/Motif-2.6B/blob/main/modeling_motif.py#L276, It looks like the `bias` should use the following fields, right?
```suggestion
            bias=getattr(config, "use_bias", False),
```

File: vllm/model_executor/models/motif.py [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+                config, "original_max_position_embeddings", None):
+            rope_scaling["original_max_position_embeddings"] = (
+                config.original_max_position_embeddings)
+        max_position_embeddings = getattr(config, "max_position_embeddings",
+                                          8192)
+        attention_bias = getattr(config, "attention_bias", False) or getattr(
+            config, "bias", False)
+        bias_o_proj = attention_bias
+        if hasattr(config, 'qkv_bias'):
+            attention_bias = config.qkv_bias
```
Comment (jeejeelee) [inline vllm/model_executor/models/motif.py]: The bias  `getattr` here also seems to have some issues

File: vllm/model_executor/models/motif.py [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+        # Fully Connected
+        hidden_states, residual = self.post_attention_layernorm(
+            hidden_states, residual)
+        hidden_states = self.mlp(hidden_states)
+        return hidden_states, residual
+
+
+# Motif model uses differential attention
+# Only supported in v0 (no chunked prefill support)
+class MotifForCausalLM(LlamaForCausalLM, SupportsV0Only):
```
Comment (jeejeelee) [inline vllm/model_executor/models/motif.py:326]: Do we need to add these [assertions](https://github.com/vllm-project/vllm/blob/v0.10.2rc1/vllm/model_executor/models/phi4flash.py#L609-L612) ?

File: vllm/model_executor/models/motif.py [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+            prefix=f"{prefix}.down_proj",
+        )
+        if hidden_act != "poly_norm":
+            raise NotImplementedError(f"Unsupported activation: {hidden_act}. "
+                                      "Only poly_norm is supported for now.")
+        self.act_fn = PolyNorm()
+        self.intermediate_size = intermediate_size
+        tp_size = get_tensor_model_parallel_world_size()
+        if hidden_act == "poly_norm" and tp_size > 1:
+            raise NotImplementedError(
```
Comment (jeejeelee) [inline vllm/model_executor/models/motif.py:70]: ~~QQ:Could you please explain why TP is not supported?~~
After taking a deeper look at the code, it should be due to `_norm`

**Comment by jeejeelee** [comment]
I also added `torch.compile` benchmark on my local A800, and the results are as follows
image

**Review by jeejeelee — APPROVED** [review]
After resolving the comments about the assertion above, overall LGTM. Thank you for your patience and contribution.
@mgoin  @yewentao256   If possible, could you please review this PR again, especially the kernel-related code? Thank you very much.

File: vllm/model_executor/models/motif.py (hunk 6 above)
Comment (ca1207) [reply vllm/model_executor/models/motif.py:326]: Thanks, review comment applied 
By the way I think [this part](https://github.com/vllm-project/vllm/blob/5438967fbc7a10ae6eee7a98182f4de94101e858/vllm/model_executor/models/phi4flash.py#L612) of the assert msg should be updated from prefix caching to chunked prefill

**Comment by ca1207** [comment]
@jeejeelee Thank you very much for your kind and detailed review. It was very helpful.

**Comment by ca1207** [comment]
@mgoin @yewentao256 just a gentle ping on this [comment](https://github.com/vllm-project/vllm/pull/23414#pullrequestreview-3176836598)
It would be great if you could review this when you get a chance.

**Comment by jeejeelee** [comment]
Please resolve the branch conflicts, then considering that the changes in this PR won't significantly affect other features,  we can go ahead and merge.

File: csrc/layernorm_kernels.cu [hunk 8]
```diff
@@ -140,6 +140,211 @@ fused_add_rms_norm_kernel(
...
+
+  float3 thread_variances = make_float3(variance, variance2, variance3);
+
+  struct SumOp {
+    __device__ float3 operator()(const float3& a, const float3& b) const {
+      return make_float3(a.x + b.x, a.y + b.y, a.z + b.z);
+    }
+  };
+
+  using BlockReduce = cub::BlockReduce<float3, 1024>;
```
Comment (yewentao256) [inline csrc/layernorm_kernels.cu:242]: Here we set BlockReduce with blockDim.x 1024
But blockDim.x is decided in

```bash
const int max_block_size = (num_tokens < 256) ? 1024 : 256;
dim3 block(std::min(hidden_size, max_block_size)); 
```

In this case, worried the inconsistency will cause trouble, could you test on this?

File: csrc/layernorm_kernels.cu [hunk 9]
```diff
@@ -219,3 +424,48 @@ void fused_add_rms_norm(torch::Tensor& input,     // [..., hidden_size]
...
+        hidden_size);                                                         \
+  });
+
+void poly_norm(torch::Tensor& out,     // [..., hidden_size]
+               torch::Tensor& input,   // [..., hidden_size]
+               torch::Tensor& weight,  // [3]
+               torch::Tensor& bias,    // [1]
+               double epsilon) {
+  TORCH_CHECK(out.is_contiguous());
+  TORCH_CHECK(input.is_contiguous());
```
Comment (yewentao256) [inline csrc/layernorm_kernels.cu:442]: Is this kernel supporting the in-place version?
If not, please add `TORCH_CHECK(out.data_ptr() != input.data_ptr())` since we have `__restrict__`

**Review by yewentao256 — COMMENTED** [review]
Thanks for the work! Overall looks good to me, just a few comments

File: csrc/layernorm_kernels.cu (hunk 8 above)
Comment (ca1207) [reply csrc/layernorm_kernels.cu:242]: Thanks for the comment. In test_layernorm.py, we already have a test case for num_tokens < 256, and it works fine in that case.
```
NUM_TOKENS = [7, 83, 4096]  # Arbitrary values for testing
HIDDEN_SIZES = [8, 768, 769, 770, 771, 5120, 5124, 5125, 5126, 8192,
                8199]  # Arbitrary values for testing
ADD_RESIDUAL = [False, True]
```

File: csrc/layernorm_kernels.cu (hunk 9 above)
Comment (ca1207) [reply csrc/layernorm_kernels.cu:442]: Thanks, applied at this commit https://github.com/vllm-project/vllm/pull/23414/commits/c9779bc73342045a1808a47e14df394a5bfb372c

File: csrc/layernorm_kernels.cu (hunk 8 above)
Comment (yewentao256) [reply csrc/layernorm_kernels.cu:242]: Looks good, thanks!

**Review by yewentao256 — APPROVED** [review]
LGTM, thanks for the work!

**Comment by DarkLight1337** [comment]
Please fix the merge conflicts

**Comment by ca1207** [comment]
@DarkLight1337 
Hi! have quick question
If the main branch has been updated without any conflicts, do I still need to keep updating it?
Currently, while my GitHub Action is running, main continues to receive updates, and each time I merge these updates, the workflow restarts from the beginning.

**Comment by DarkLight1337** [comment]
No need to update as long as there are no merge conflicts.
